external_services/lti1p3.py

```diff
@@ -0,0 +1,370 @@
+import json
+import logging
+import datetime
+from typing import List, Dict, Any, Tuple
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.core.exceptions import PermissionDenied
+from django.urls import reverse
+from aplus.api import api_reverse
+
+import jwt
+from jwcrypto import jwk
+from oauthlib.openid import RequestValidator as OIDCRequestValidator
+from oauthlib.common import Request as OAuthRequest
+from oauthlib.oauth2.rfc6749.errors import OAuth2Error
+
+from lib.helpers import build_aplus_url
+from exercise.exercise_models import LTI1p3Exercise
+from course.models import CourseInstance
+from .models import LTI1p3Service
+from authorization.models import OAuth2BearerToken
+
+
+logger = logging.getLogger('aplus.external_services')
+
+def _add_claim(id_token: dict, key: str, value: Any) -> None:
+    id_token.update({f"https://purl.imsglobal.org/spec/lti/claim/{key}": value})
+
+
+def prepare_lti1p3_initiate_login(
+        service: LTI1p3Service,
+        user: User,
+        instance: CourseInstance,
+        exercise: LTI1p3Exercise = None
+        ) -> List[Tuple[str, str]]:
+    """
+    Prepares LTI 1.3 initiate login request sent from A+ to LTI Tool, which is expected
+    to respond by sending an Authentication request back. Returns parameters included in
+    the request.
+    """
+    #def __init__(self, service: LTI1p3Service, user: User, instance: CourseInstance, exercise: LTI1p3Exercise = None):
+    if user.is_anonymous:
+        raise PermissionDenied()
+
+    # LTI message hint is used to find course instance / exercise at later stages of handshake
+    if exercise:
+        hint = exercise.get_resource_link_id()
+    else:
+        hint = str(instance.pk)
+
+    parameters = [
+        ("iss", build_aplus_url('')),
+        ("target_link_uri", service.url),
+        ("login_hint", str(user.pk)),
+        ("client_id", service.client_id),
+        ("lti_deployment_id", service.deployment_id),
+        ("lti_message_hint", hint),
+    ]
+    return parameters
+
+
+class LTI1p3Client:
+    """
+    Represents a validated OAuth client (LTI 1.3 tool) interacting with A+ LTI platform.
+    """
+    def __init__(self, client_id):
+        self.client_id = client_id
+
+
+class LTI1p3AuthValidator(OIDCRequestValidator):
+    """
+    Implementation of oauthlib Auhtentication request validator for incoming OAuth2
+    requests used in LTI 1.3.
+
+    @param user: Django user associated with HTTP requests, when used in login
+    authorization request triggered by A+ user.
+    None, if request has originated from LTI tool.
+    
+    @param params: URL query parameters from OAuth request.
+    """
+    def __init__(self, user: User, params: dict, message_hint: str):
```

Comment:
`message_hint` type should now be `Optional[str]` after the changes to `LTI1p3AuthRequestView`